gradle-guide-error-prone/src/main/java/com/palantir/gradle/guide/errorprone/utils/MethodCallGraph.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.guide.errorprone.utils;
+
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.ImmutableGraph;
+import com.google.common.graph.MutableGraph;
+import com.google.common.graph.Traverser;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.CompilationUnitTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.util.TreePathScanner;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * A directed graph of "who is called by who" within this compilation unit (source file).
+ */
+public class MethodCallGraph {
+    /**
+     * Nodes are {@code MethodSymbol}s. There is an edge from {@code f} to {@code g} iff {@code f} is called by
+     * {@code g}.
+     */
+    private final ImmutableGraph<MethodSymbol> callGraph;
+
+    public MethodCallGraph(CompilationUnitTree compilationUnitTree) {
```

Comment:
We should make the constructor private so that we _force_ bugcheckers to get the cached versioned - no chance they can make new ones of these that aren't shared.

Reply:
Yep. I wasn't sure if the `MethodCallGraph` should concern itself with caching but this is a good way to prevent perf regressions.